src/mrinufft/trajectories/densities.py

```diff
@@ -0,0 +1,134 @@
+
+    if method == "random":
+        choices = rng.choice(
+            np.arange(max_nb_samples),
+            size=nb_samples,
+            p=density.flatten(),
+            replace=False,
+        )
+        locations = np.indices(shape).reshape((nb_dims, -1))[:, choices]
+        locations = locations.T
+        locations = 2 * KMAX * locations / np.max(locations) - KMAX
+    elif method == "lloyd":
+        kmeans = (
+            KMeans(n_clusters=nb_samples)
+            if nb_dims <= 2
+            else BisectingKMeans(n_clusters=nb_samples)
+        )
+        kmeans.fit(
+            np.indices(density.shape).reshape((nb_dims, -1)).T,
+            sample_weight=density.flatten(),
+        )
+        locations = kmeans.cluster_centers_ - np.array(density.shape) / 2
+        locations = KMAX * locations / np.max(np.abs(locations))
+    else:
+        raise ValueError(f"Unknown sampling method {method}.")
+    return locations
+
+
+def create_cutoff_decay_density(shape, cutoff, decay, resolution=None):
+    shape = np.array(shape)
+    nb_dims = len(shape)
+
+    if not resolution:
+        resolution = np.ones(nb_dims)
+
+    distances = np.indices(shape).astype(float)
+    for i in range(nb_dims):
+        distances[i] = distances[i] + 0.5 - shape[i] / 2
+        distances[i] = distances[i] / shape[i] * resolution[i]
+    distances = nl.norm(distances, axis=0)
+
+    cutoff = cutoff * np.max(distances) if cutoff else np.min(distances)
+    density = np.ones(shape)
+    decay_mask = np.where(distances > cutoff, True, False)
+
+    density[decay_mask] = (cutoff / distances[decay_mask]) ** decay
+    density = density / np.sum(density)
+    return density
+
+
+def create_polynomial_density(shape, decay, resolution=None):
+    return create_cutoff_decay_density(
+        shape, cutoff=0, decay=decay, resolution=resolution
+    )
+
+
+def create_energy_density(dataset):
+    nb_dims = len(dataset.shape) - 1
+    axes = range(1, nb_dims + 1)
+
+    kspace = nf.fftshift(nf.fftn(nf.fftshift(dataset, axes=axes), axes=axes), axes=axes)
+    energy = np.mean(np.abs(kspace), axis=0)
+    density = energy / np.sum(energy)
+    return density
+
+
+def create_chauffert_density(shape, wavelet_basis, nb_wavelet_scales, verbose=False):
+    nb_dims = len(shape)
+    indices = np.indices(shape).reshape((nb_dims, -1)).T
+
+    density = np.zeros(shape)
+    unit_vector = np.zeros(shape)
+    for ids in tqdm(indices, disable=not verbose):
+        ids = tuple(ids)
+        unit_vector[ids] = 1
+        fourier_vector = nf.ifftn(unit_vector)
+        coeffs = pw.wavedecn(
+            fourier_vector, wavelet=wavelet_basis, level=nb_wavelet_scales
+        )
+        coeffs, _ = pw.coeffs_to_array(coeffs)
+        density[ids] = np.max(np.abs(coeffs)) ** 2
+        unit_vector[ids] = 0
+
+    density = density / np.sum(density)
+    return nf.ifftshift(density)
+
+
+def create_fast_chauffert_density(shape, wavelet_basis, nb_wavelet_scales):
```

Comment:
To keep in mind: https://arxiv.org/pdf/1910.14513, some other interesting results with Anna.

Reply:
Is there something specific in the paper you would like to highlight ? Should we open an issue to create some extension in the future ?

Reply:
Also @chaithyagr I note in the paper figure that there is a TWIRL implementation sleeping somewhere in our cardboards. Do you know where it is so it can be added to the package ?

Reply:
Isnt TWIRL just TPI in 2D? It should be fairly straightforward from our codes.

Reply:
> Is there something specific in the paper you would like to highlight ? Should we open an issue to create some extension in the future ?

I think @philouc can comment better than me. I just felt its good to add in docs, Its been long having gone through this work, but from what I remember its extension of the earlier work for anisotropic sampling, which can be of interest too.. But surely not to be pursued in this work